### PR TITLE
[DO NOT MERGE] Forgive me -- Double Memory 

### PR DESCRIPTION
**Pull request by ruchim**
May I get a review on the design, but not the scala? 

Sorry it's so ugly, my apologies.

Thanks!

### CODE REVIEW

**Comment by cjllanwarne** [comment]
This seems to be interestingly connected with the spec change https://github.com/openwdl/wdl/pull/315 (so cc @patmagee)

As that PR is currently written, we would be fine to do the scheme like this, because the `memory` section says "you can provide as much memory as you want, as long as it's over this amount", but it feels like we're in danger of writing non-portable WDLs like this because the incentive is to write a small value first and rely on the doubling to catch you if necessary.

FWIW I'd rather go down the route of:
- `memory` is treated as the "guaranteed to work" ceiling amount
- We could start by having a much lower `memory_to_try_first` attribute representing the first value to try
- If the task fails, we can then double it from the low baseline, until either the task succeeds or we reach the `memory` ceiling, and at that point we don't try any further.

Does that make sense?

**Comment by ruchim** [comment]
Hey @cjllanwarne,

That's an interesting point. When it comes to resources (like CPU/Memory/disk) I think the only incentive a user has is to save on cost -- and they do so by having the default values for resources be enough to cover the "average" use case. I don't think this change shifts the incentive since if people start lowering the min memory required, that would mean on average tasks will retry more often, leading to greater spend. Though I'll let @ldgauthier speak from the WDL author perspective!

**Comment by ldgauthier** [comment]
There are some cases where I want the task to succeed no matter the cost/memory.  (e.g. I don't know how much memory it will take because I've never processed that many samples together before.) If there was a way to set infinite max memory then that could work, but something about that gives me a bad feeling in my stomach.

**Comment by ldgauthier** [comment]
Maybe I'm being melodramatic. Probably if something needs 128GB of memory then something has gone wrong it we need a new method. It's worth a failure to go take a look.  Yeah, I could be in board we two parameters like the Java xmx/xms

**Comment by cjllanwarne** [comment]
I guess what we call these attributes doesn't really matter so much as, we have to accept one of these worlds:
* The spec defines a maximum memory value, above which Cromwell will never go
* The spec makes it clear that memory increasing will sometimes be required to complete tasks when they specify "average" `memory` attributes
* Cromwell will be able to run WDLs which will not run anywhere else... and thus we would have to be very strict in policing our "best practices" WDL to makes sure it can be run on other engines

**Comment by rhpvorderman** [comment]
> May I get a review on the design, but not the scala?

Wow thanks for adding this. This is exactly what we need on our cluster. Sometimes the scheduler aggressively kills jobs based on VMEM (instead of actual mem). So retrying with upping the memory requirements is a nice way to circumvent this annoying issue. (Instead of using insane memory requirements to make sure it passes in 99% of the cases).

As for the design. I would add
1. A number of attempts configurable parameter in the config. A sane default would be 1. Meaning that this feature will not be used by default, for reasons elaborated on later.
2. A factor with which the memory is increased on each attempt. So if the factor is 1.5 > Attempt 1 will be 1.5^0 = 1 times the memory, Attempt 2 1.5^1 = 1.5, Attempt 3 = 1.5^2 = 2.25. A sane default here would be 2 I guess.

As for @cjllanwarne's concerns:

>  The spec defines a maximum memory value, above which Cromwell will never go
 
I think the spec just states the value that should be given to whatever backend. But semantics aside, I guess that means the same as saying it is the maximum.

> Cromwell will be able to run WDLs which will not run anywhere else... and thus we would have to be very strict in policing our "best practices" WDL to makes sure it can be run on other engines

This is a very good reason to not enable this feature by default. But since there are very good reasons for having this feature, having it as a configurable option will be very very nice. Let the user decide how they want to treat their memory requirements. That is the most user-friendly way to do. This is way I think a sane default for the number of attempts should be 1 (i.e. no retries).

**Comment by cjllanwarne** [comment]
The opinions from other engine developers in the WDL spec makes it sound like they're quite relaxed about this doubling scheme, so IMO it's fine to go ahead with this as-is.

Depending on whether I get my way or not, we might or might not see a spec change in WDL 2.0 with a maximum memory requirement above which Cromwell should not go, but that would be a bridge for us to cross in the future.

TL;DR: 👍

**Comment by geoffjentry** [comment]
I think this is fine, re the incoming WDL spec changes. We also know that there'll soon be (in some shape/way/form) a codified mechanism to override runtime attributes. One could think of this as an automated override

**Comment by ruchim** [comment]
Closing as I have some feedback to make changes!

**Comment by dinvlad** [comment]
It would be a really nice to have feature indeed!

In addition to what was said above, we're also planning to work on runtime metrics reporting at DSP hackathon with @mohawkTrail and @rexwangcc . In the future, the "automatic retry" feature could enable us to collect runtime statistics much more easily (in a fully automated fashion). From this, we could then build accurate models on how much resources are actually needed, for a given set of inputs (so we could predict the amount of resources for future workflows, instead of relying on guesswork or retries).

Without this feature, we have to either overshoot the amount of resources uniformly for all tasks (and then collect how much they actually need), or somehow retry this in a "ladder" fashion using an automated script that does a "parameter sweep" that we then pass to our workflow as inputs (or directly edit using something like MiniWDL perhaps?).
